File: internal/cmds/intel.go
package cmds

import (
	"context"
	"errors"
	"fmt"
	"math"
	"strings"
	"time"

	"github.com/rprtr258/twitch-bot/internal/message"
	"github.com/rprtr258/twitch-bot/internal/permissions"
	"github.com/rprtr258/twitch-bot/internal/services"

	"github.com/nicklaw5/helix/v2"
)

func formatDuration(d time.Time) string {
	var parts []string
	now := time.Now()

	{
		var years int
		for d.AddDate(1, 0, 0).Before(now) {
			years++
			d = d.AddDate(1, 0, 0)
		}
		if years > 0 {
			parts = append(parts, fmt.Sprintf("%dy", years))
		}
	}

	{
		var months int
		for d.AddDate(0, 1, 0).Before(now) {
			months++
			d = d.AddDate(0, 1, 0)
		}
		if months > 0 {
			parts = append(parts, fmt.Sprintf("%dmo", months))
		}
	}

	{
		var days int
		for d.AddDate(0, 0, 1).Before(now) {
			days++
			d = d.AddDate(0, 0, 1)
		}
		if days > 0 {
			parts = append(parts, fmt.Sprintf("%dd", days))
		}
	}

	minutes := (int64)(math.Floor(now.Sub(d).Minutes()))
	if minutes/60 > 0 {
		parts = append(parts, fmt.Sprintf("%dh", minutes/60))
	}

	parts = append(parts, fmt.Sprintf("%dm", minutes%60))

	return strings.Join(parts, "")
}

type IntelCmd struct{}

func (IntelCmd) RequiredPermissions() []string {
	return []string{}
}

func (IntelCmd) Command() string {
	return "?intel"
}

func (IntelCmd) Description() string {
	return "Gather intel on user"
}

func (IntelCmd) Run(ctx context.Context, s *services.Services, perms permissions.PermissionsList, msg message.TwitchMessage) (string, error) {
	words := strings.Split(msg.Text, " ")
	if len(words) < 2 {
		return "No username provided", nil
	}

	login := words[1]
	resp, err := s.TwitchApiClient.GetUsers(&helix.UsersParams{
		Logins: []string{login},
	})
	if err != nil {
		return "", err
	}
	if resp.ErrorMessage != "" {
		return "", errors.New(resp.ErrorMessage)
	}

	users := resp.Data.Users
	if len(users) == 0 {
		return fmt.Sprintf("%s not found", login), nil
	}

	user := users[0]
	createdAt := user.CreatedAt.Time
	delta := formatDuration(createdAt)
	return fmt.Sprintf(
		"created=%s ago=%s broadcaster_type=%q id=%s login=%q display_name=%q view_count=%d",
		createdAt.Format("15:04.2.1.2006"),
		delta,
		user.BroadcasterType,
		user.ID,
		user.Login,
		user.DisplayName,
		user.ViewCount,
	), nil
}

File: internal/cmds/intel.ts
import {type Command} from "../cmds.ts";
import * as message from "../message.ts";
import * as permissions from "../permissions.ts";
import * as services from "../services.ts";

function formatDuration(d: Date): string {
  const now = new Date();

  let parts: string[] = [];
  {
    const years = d.getFullYear() - now.getFullYear();
    if (years > 0)
      parts.push(`${years}y`);
    d.setFullYear(now.getFullYear());
  }

  {
    const months = d.getMonth() - now.getMonth();
    if (months > 0)
      parts.push(`${months}mo`);
    d.setMonth(now.getMonth());
  }

  {
    const days = d.getDate() - now.getDate();
    if (days > 0)
      parts.push(`${days}d`);
    d.setDate(now.getDate());
  }

  const minutes = now.getMinutes() - d.getMinutes();
  if (Math.floor(minutes/60) > 0) {
    parts.push(`${Math.floor(minutes/60)}h`);
  }

  parts.push(`${minutes%60}m`);

  return parts.join("");
}

export default {
  RequiredPermissions: [],
  Command: "?intel",
  Description: "Gather intel on user",
  Run(s: services.Services, perms: permissions.PermissionsList, msg: message.TwitchMessage): string {
    const words = msg.Text.split(" ");
    if (words.length < 2) {
      return "No username provided"
    }

    const login = words[1];
    const resp = s.TwitchApiClient.GetUsers({Logins: [login]});
    if (resp.ErrorMessage !== "") {
      throw new Error(resp.ErrorMessage);
    }

    const users = resp.Data.Users;
    if (users.length === 0) {
      return `${login} not found`;
    }

    const user = users[0];
    const createdAt = user.CreatedAt.Time;
    const delta = formatDuration(createdAt);
    return [
      `created=${createdAt.Format("15:04.2.1.2006")}`,
      `ago=${delta}`,
      `broadcaster_type=${user.BroadcasterType}`,
      `id=${user.ID}`,
      `login=${user.Login}`,
      `display_name=${user.DisplayName}`,
      `view_count=${user.ViewCount}`,
    ].join(" ");
  },
} as Command;
